File: Logic/BookDM.cs
using AmazonIntegration;
using EntityModels;
using Infrastructure.Data;
using Infrastructure.Logic;
using Newtonsoft.Json;
using Shared.Interfaces;
using Shared.Services;
using System.Configuration;
using System.Linq;
using ViewModels;

namespace Logic
{
    public class BookDM : BaseDM, IBookDM
    {
        public BookDM(IFactory factory) : base(factory)
        {

        }

        public void Delete(long id)
        {
            using (var bookRepo = Factory.GetService<IBookRepo>())
            {
                bookRepo.Delete(id);
            }
        }

        public void Update(BookVM model)
        {
            var book = MapperService.Map<BookEM>(model);
            using (var bookRepo = Factory.GetService<IBookRepo>())
            using (var scope = new TransactionService())
            {
                long id;

                id = book.Id.Value;
                bookRepo.Update(book);
                var authorsIds = model.Authors.Select(a => a.Id.Value);
                bookRepo.UpdateAuthors(id, authorsIds);

                scope.Complete();
            }
        }

        public void Create(BookVM model)
        {
            var book = MapperService.Map<BookEM>(model);
            using (var bookRepo = Factory.GetService<IBookRepo>())
            using (var scope = new TransactionService())
            {
                long id;

                id = bookRepo.Create(book);
                var authorsIds = model.Authors.Select(a => a.Id.Value);
                bookRepo.UpdateAuthors(id, authorsIds);

                scope.Complete();
            }
        }

        public BookVM Get(long? id)
        {
            using (var bookRepo = Factory.GetService<IBookRepo>())
            {
                if (id.HasValue)
                {
                    var bookEM = bookRepo.Get(id.Value);
                    var bookVM = MapperService.Map<BookVM>(bookEM);
                    return bookVM;
                }
                return null;
            }
        }

        public DataTableResponseVM<BookVM> GetList(DataTableRequestVM model)
        {
            var dataTableEM = MapperService.Map<DataTableRequestEM>(model);

            using (var bookRepo = Factory.GetService<IBookRepo>())
            {
                var responseEM = bookRepo.GetList(dataTableEM);
                var responseVM = MapperService.Map<DataTableResponseVM<BookVM>>(responseEM);
                responseVM.draw = model.Draw;

                return responseVM;
            }
            
        }

        public string Publish(BookVM model)
        {
            var em = MapperService.Map<BookEM>(model);
            var json = JsonConvert.SerializeObject(em);
            var arn = ConfigurationManager.AppSettings["AWSSNSTopicARN"];
            var sns = new SNS();
            return sns.PublishEntity(arn, json, "Book", "Book");
        }
    }
}

File: LogicVB/BookDM.vb
Imports EntityModels
Imports Infrastructure.Data
Imports Infrastructure.Logic
Imports [Shared].Interfaces
Imports [Shared].Services
Imports ViewModels

Public Class BookDM
    Inherits BaseDM
    Implements IBookDM

    Public Sub New(factory As IFactory)
        MyBase.New(factory)
    End Sub

    Public Sub Delete(id As Long) Implements IBookDM.Delete
        Using bookRepo = Factory.GetService(Of IBookRepo)()
            bookRepo.Delete(id)
        End Using
    End Sub

    Public Sub Update(model As BookVM) Implements IBookDM.Update
        Dim book = MapperService.Map(Of BookEM)(model)
        Using bookRepo = Factory.GetService(Of IBookRepo)(),
                    scope = New TransactionService()

            Dim id As Long

            id = book.Id.Value
            bookRepo.Update(book)
            Dim authorsIds = model.Authors.Select(Function(a) a.Id.Value)
            bookRepo.UpdateAuthors(id, authorsIds)

            scope.Complete()
        End Using

    End Sub

    Public Sub Create(model As BookVM) Implements IBookDM.Create
        Dim book = MapperService.Map(Of BookEM)(model)
        Using bookRepo = Factory.GetService(Of IBookRepo)(),
                    scope = New TransactionService()

            Dim id As Long

            id = bookRepo.Create(book)
            Dim authorsIds = model.Authors.Select(Function(a) a.Id.Value)
            bookRepo.UpdateAuthors(id, authorsIds)

            scope.Complete()
        End Using

    End Sub

    Public Function [Get](id As Long?) As BookVM Implements IBookDM.Get
        Using bookRepo = Factory.GetService(Of IBookRepo)()

            If (id.HasValue) Then
                Dim BookEM = bookRepo.Get(id.Value)
                Dim BookVM = MapperService.Map(Of BookVM)(BookEM)
                Return BookVM
            End If
            Return Nothing
        End Using

    End Function

    Public Function GetList(model As DataTableRequestVM) As DataTableResponseVM(Of BookVM) Implements IBookDM.GetList
        Dim dataTableEM = MapperService.Map(Of DataTableRequestEM)(model)

        Using bookRepo = Factory.GetService(Of IBookRepo)()
            Dim responseEM = bookRepo.GetList(dataTableEM)
            Dim responseVM = MapperService.Map(Of DataTableResponseVM(Of BookVM))(responseEM)
            responseVM.draw = model.Draw

            Return responseVM
        End Using
    End Function
End Class

